Fixes SI-8689 by delegating Promise.completeWith to Promise.tryCompleteW...

viktorklang opened the pull request:
...ith

as well as deprecates `Promise.completeWith` since it is now redundant.

Backport to 2.11.x and 2.10.x without the deprecation

retronym commented:
@viktorklang: Thanks for picking this up!

Could you please reformat the commit messages along the lines of the following?
We like them to fix the git conventions of a 70 char subject and body line wrapped
to 78 chars, and to describe the problem, solution.

```
SI-8689 Avoid internal error in Promise after sequence of completions

A sequence of calls to ... and ... could result in an IllegalStateException.

This happened because ...

This commit avoids problem by delegating Promise.completeWith to Promise.tryComplete
which has the effect of ...

It has been tested by ...
```

```
SI-8689 Deprecates Promise.completeWith in favor of Promise.tryCompleteWith

As seen from the bug in the previous commit, this method is ....

It is now redundant given ..., so we can deprecate it.
```

It would also make our lives easier if the main bug fix could be submitted to the 2.10.x branch directly. We will forward merge this to 2.11.x and 2.12.x as a matter of course. This workflow is a bit easier for us than backporting.

Thanks!

viktorklang commented:
@retronym I've updated the commit messages, but won't have time to repoint the target. See if it merges cleanly or not. I'm on holiday for about 2 weeks.

viktorklang committed:
SI-8689 Avoid internal error in Promise after sequence of completions

Calling `completeWith` when the `DefaultPromise` is already completed,
leads to callbacks not being properly executed.

This happened because `Future.InternalCallbackExecutor` extends
`BatchingExecutor` which assumes `unbatchedExecute` to be async,
when in this case it is sync, and if there is an exception thrown
by executing the batch, it creates a new batch with the remaining
items from the current batch and submits that to `unbatchedExecute`
and then rethrows, but if you have a sync `unbatchedExecute`, it will
fail since it is not reentrant, as witnessed by the failed `require`
as reported in this issue.

This commit avoids problem by delegating `completeWith` to `tryComplete`,
which has the effect of using `onComplete` + `tryComplete` i.s.o. `complete`,
which means that when it fails (because of a benign race condition between
completers) it won't throw an exception.

It has been tested by the minimized reproducer.

retronym commented:
I've submitted this against 2.10.x as #4289.
